File: backend/src/common/metrics.ts
import {
  CallHandler,
  ExecutionContext,
  Injectable,
  NestInterceptor,
} from '@nestjs/common';
import type { Request, Response } from 'express';
import { Observable, finalize } from 'rxjs';

interface RouteStats {
  count: number;
  errors5xx: number;
  durations: number[];
}

const WINDOW = 500;

/** Métricas básicas en memoria por ruta: solicitudes, errores 5xx, latencia promedio y p95. */
@Injectable()
export class MetricsRegistry {
  readonly startedAt = Date.now();
  private readonly routes = new Map<string, RouteStats>();

  observe(route: string, statusCode: number, durationMs: number): void {
    const stats = this.routes.get(route) ?? {
      count: 0,
      errors5xx: 0,
      durations: [],
    };
    stats.count += 1;
    if (statusCode >= 500) stats.errors5xx += 1;
    stats.durations.push(durationMs);
    if (stats.durations.length > WINDOW) stats.durations.shift();
    this.routes.set(route, stats);
  }

  snapshot() {
    const routes: Record<
      string,
      { count: number; errors5xx: number; avgMs: number; p95Ms: number }
    > = {};
    for (const [route, stats] of [...this.routes.entries()].sort(([a], [b]) =>
      a.localeCompare(b),
    )) {
      const sorted = [...stats.durations].sort((a, b) => a - b);
      const p95 = sorted[Math.max(0, Math.ceil(sorted.length * 0.95) - 1)] ?? 0;
      const avg = sorted.length
        ? sorted.reduce((sum, d) => sum + d, 0) / sorted.length
        : 0;
      routes[route] = {
        count: stats.count,
        errors5xx: stats.errors5xx,
        avgMs: Math.round(avg * 100) / 100,
        p95Ms: Math.round(p95 * 100) / 100,
      };
    }
    return {
      uptimeSeconds: Math.round((Date.now() - this.startedAt) / 1000),
      routes,
    };
  }
}

@Injectable()
export class MetricsInterceptor implements NestInterceptor {
  constructor(private readonly registry: MetricsRegistry) {}

  intercept(context: ExecutionContext, next: CallHandler): Observable<unknown> {
    if (context.getType() !== 'http') return next.handle();
    const http = context.switchToHttp();
    const request = http.getRequest<Request>();
    const response = http.getResponse<Response>();
    const started = performance.now();
    const route = `${request.method} ${(request.route as { path?: string } | undefined)?.path ?? request.path}`;

    return next.handle().pipe(
      finalize(() => {
        const duration = performance.now() - started;
        this.registry.observe(route, response.statusCode, duration);
        response.setHeader('X-Response-Time-Ms', duration.toFixed(1));
      }),
    );
  }
}

File: python-service/app/metrics.py
"""Métricas básicas en memoria: conteo, errores y latencias por ruta."""

import math
import time
from collections import defaultdict, deque
from threading import Lock
from typing import Any


class MetricsRegistry:
    """Ventana deslizante de latencias por ruta (p. ej. `POST /v1/recommendations/rank`)."""

    def __init__(self, window: int = 500) -> None:
        self._window = window
        self._lock = Lock()
        self._durations: dict[str, deque[float]] = defaultdict(lambda: deque(maxlen=self._window))
        self._counts: dict[str, int] = defaultdict(int)
        self._errors: dict[str, int] = defaultdict(int)
        self.started_at = time.time()

    def observe(self, route: str, status_code: int, duration_ms: float) -> None:
        with self._lock:
            self._counts[route] += 1
            if status_code >= 500:
                self._errors[route] += 1
            self._durations[route].append(duration_ms)

    def snapshot(self) -> dict[str, Any]:
        with self._lock:
            routes = {}
            for route, count in sorted(self._counts.items()):
                durations = sorted(self._durations[route])
                p95_index = max(0, math.ceil(0.95 * len(durations)) - 1)
                routes[route] = {
                    "count": count,
                    "errors5xx": self._errors[route],
                    "avgMs": round(sum(durations) / len(durations), 2) if durations else 0.0,
                    "p95Ms": round(durations[p95_index], 2) if durations else 0.0,
                }
            return {"uptimeSeconds": round(time.time() - self.started_at, 1), "routes": routes}

    def reset(self) -> None:
        with self._lock:
            self._durations.clear()
            self._counts.clear()
            self._errors.clear()


metrics = MetricsRegistry()
